File: superstitioMain/java/superstitio/cards/colorless/FindCardAndHang.java
package superstitio.cards.colorless;

import basemod.cardmods.ExhaustMod;
import basemod.helpers.CardModifierManager;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.characters.AbstractPlayer;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import superstitio.DataManager;
import superstitio.cards.SuperstitioCard;
import superstitioapi.actions.AutoDoneInstantAction;
import superstitioapi.actions.ChoseCardFromGridSelectWindowAction;
import superstitioapi.hangUpCard.CardOrb_WaitCardTrigger;

public class FindCardAndHang extends SuperstitioCard {
    public static final String ID = DataManager.MakeTextID(FindCardAndHang.class);

    public static final CardType CARD_TYPE = CardType.SKILL;

    public static final CardRarity CARD_RARITY = CardRarity.RARE;

    public static final CardTarget CARD_TARGET = CardTarget.SELF;

    private static final int COST = 1;
    private static final int MAGIC = 10;
    private static final int UPGRADE_MAGIC = -4;
    private static final int CHOSE_CARD = 10;
    private static final int WAIT_TIME = 3;

    public FindCardAndHang() {
        super(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET, CardColor.COLORLESS, "special");
        this.setupMagicNumber(MAGIC, UPGRADE_MAGIC);
        CardModifierManager.addModifier(this, new ExhaustMod());
    }

    private void HangUpSpecificCard(AbstractCard card) {
        AbstractCard copyCard = card.makeStatEquivalentCopy();
        copyCard.exhaust = true;
        AbstractCard showUpCard = this.makeStatEquivalentCopy();
        showUpCard.cardsToPreview = copyCard;
        AutoDoneInstantAction.addToBotAbstract(() -> {
            AbstractDungeon.player.drawPile.removeCard(card);
            new CardOrb_WaitCardTrigger(card, AbstractDungeon.player.hand, magicNumber, (orb, usedcard) -> {
            })
                    .setShowCard(showUpCard)
                    .addToBot_HangCard();
        });
    }

    @Override
    public void use(AbstractPlayer player, AbstractMonster monster) {
        addToBot(new ChoseCardFromGridSelectWindowAction(AbstractDungeon.player.drawPile,
                this::HangUpSpecificCard)
                .setWindowText(String.format(getEXTENDED_DESCRIPTION()[0], CHOSE_CARD))
                .setChoseAmount(CHOSE_CARD)
                .setAnyNumber(true)
        );
    }

    @Override
    public void upgradeAuto() {
    }
}

File: superstitioMain/java/superstitio/cards/colorless/FindCardAndHang.kt
package superstitio.cards.colorless

import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.characters.AbstractPlayer
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.monsters.AbstractMonster
import superstitio.DataManager
import superstitio.cards.SuperstitioCard
import superstitioapi.actions.AutoDoneInstantAction
import superstitioapi.actions.ChoseCardFromGridSelectWindowAction
import superstitioapi.cards.addExhaustMod
import superstitioapi.hangUpCard.CardOrb_CardTrigger
import superstitioapi.hangUpCard.CardOrb_WaitCardTrigger
import superstitioapi.utils.CostSmart

class FindCardAndHang : SuperstitioCard(ID, CARD_TYPE, COST, CARD_RARITY, CARD_TARGET, CardColor.COLORLESS, "special")
{
    init
    {
        this.setupMagicNumber(MAGIC, UPGRADE_MAGIC)
        this.addExhaustMod()
    }

    private fun HangUpSpecificCard(card: AbstractCard)
    {
        val copyCard = card.makeStatEquivalentCopy()
        copyCard.exhaust = true
        val showUpCard = card.makeStatEquivalentCopy()
        showUpCard.cardsToPreview = copyCard
        AutoDoneInstantAction.addToBotAbstract {
            AbstractDungeon.player.drawPile.removeCard(card)
            CardOrb_WaitCardTrigger(
                card,
                AbstractDungeon.player.hand,
                CostSmart(magicNumber)
            ) { orb: CardOrb_CardTrigger?, usedcard: AbstractCard? -> }
                .setShowCard(showUpCard)
                .setCardRawDescriptionWillShow(cardStrings.getEXTENDED_DESCRIPTION(1))
                .addToBot_HangCard()
        }
    }

    override fun use(player: AbstractPlayer?, monster: AbstractMonster?)
    {
        ChoseCardFromGridSelectWindowAction(AbstractDungeon.player.drawPile, this::HangUpSpecificCard)
            .setWindowText(String.format(cardStrings.getEXTENDED_DESCRIPTION(0), CHOSE_CARD))
            .setChoseAmount(CHOSE_CARD)
            .setAnyNumber(true)
            .addToBot()
    }

    override fun upgradeAuto()
    {
    }

    companion object
    {
        val ID: String = DataManager.MakeTextID(FindCardAndHang::class.java)

        val CARD_TYPE: CardType = CardType.SKILL

        val CARD_RARITY: CardRarity = CardRarity.RARE

        val CARD_TARGET: CardTarget = CardTarget.SELF

        private const val COST = 1
        private const val MAGIC = 10
        private const val UPGRADE_MAGIC = -4
        private const val CHOSE_CARD = 10
        private const val WAIT_TIME = 3
    }
}
